File: experimenter.py
from SolveState import SolveState
from typing import List, Tuple
import utils
from sudoku_solver import SudokuSolver
from copy import deepcopy

def solve_sudoku(board: List[List[str]]) -> List[List[str]]:
	matrix = utils.populate_options(board)
	possible_solutions = orchestrate_solution(matrix)
	print(f'Number of possible solutions: {len(possible_solutions)}')
	if len(possible_solutions) == 0:
		print ("Board cannot be solved")
		return None
	solved_boards = list()
	for solved_matrix, solution_state, seq_steps in possible_solutions:
		if solution_state == SolveState.SOLVED:
			print("Number of steps: ", len(seq_steps))
			utils.print_steps(seq_steps)
			print(".-"*15)
			solved_boards.append(utils.convert_matrix_to_board(solved_matrix))
	return solved_boards

def orchestrate_solution(matrix: List[List[List[int]]]) -> List[Tuple[List[List[List[int]]], SolveState, List[Tuple[Tuple[int, int], int, str]]]]:
	board_solver = SudokuSolver(matrix)
	solved_matrix = board_solver.solve()
	print(board_solver.state)

	if board_solver.state == SolveState.SOLVED:
		return [(solved_matrix, board_solver.state, board_solver.seq_steps)]
	elif board_solver.state == SolveState.UNSOLVABLE:
		return []
	else:
		n, r, c = find_minimum_options(solved_matrix)
		print(n, r, c)
		print(solved_matrix[r][c])
		possible_solutions = experiment_values(solved_matrix, r, c)
		possible_complete_solutions = list()
		for e_solved_matrix, solution_state, seq_steps in possible_solutions:
			possible_complete_solutions.append((e_solved_matrix, solution_state, board_solver.seq_steps + seq_steps))
		return possible_complete_solutions

def find_minimum_options(m: List[List[List[int]]]) -> Tuple[int, int, int]:
	min_count = 9
	r, c = -1, -1
	i, j = 0, 0
	for row in m:
		for cell in row:
			if len(cell) != 1 and len(cell) < min_count:
				min_count = len(cell)
				r, c = i, j
			j += 1
		i += 1
		j = 0
	return min_count, r, c

def experiment_values(matrix: List[List[List[int]]], row_index: int, column_index: int) -> List[Tuple[List[List[List[int]]], SolveState, List[Tuple[Tuple[int, int], int, str]]]]:
	options = matrix[row_index][column_index]
	matrices = list()
	for option in options:
		print("Trying " + str(option) + " at (" + str(row_index) + "," + str(column_index) + ")")
		exp_step = ((row_index+1, column_index+1), option, "Experimental")
		temp_m = deepcopy(matrix)
		temp_m[row_index][column_index] = [option]
		possible_solutions = orchestrate_solution(temp_m)
		for sm, ss, steps in possible_solutions:
			steps = [exp_step] + steps
			matrices.append((sm, ss, steps))

	solved_matrices = list()
	for sm, ss, steps in matrices:
		if ss == SolveState.SOLVED:
			solved_matrices.append((sm, ss, steps))
		else:
			print(sm)
			assert(ss == SolveState.UNSOLVABLE)
	if len(solved_matrices) == 0:
		return []
	else:
		return solved_matrices

File: chrome_extension/src/experimenter.js
import SolveState from './SolveState.js';
import { RuleSolver } from './ruleSolver.js';
import { populateOptions, printSteps, convertMatrixToBoard } from './utils.js';

/**
 *
 * @param {number[][]} board Current board
 * @returns {{solvedBoard: number[][], steps: }[]} A list of solved boards
 */
export function solveSudoku(board) {
	const matrix = populateOptions(board);
	const possibleSolutions = orchestrateSolution(matrix);
	console.log(`Number of possible solutions: ${possibleSolutions.length}`);
	if (possibleSolutions.length === 0) {
		console.log('Board cannot be solved');
		return null;
	}
	console.dir(possibleSolutions);

	const solvedBoardsWithSteps = [];
	for (const [solvedMatrix, solutionState, seqSteps] of possibleSolutions) {
		if (solutionState === SolveState.SOLVED) {
			console.log('Number of steps: ', seqSteps.length);
			printSteps(seqSteps);
			console.log('-.'.repeat(15));
			solvedBoardsWithSteps.push({
				solvedBoard: convertMatrixToBoard(solvedMatrix),
				steps: seqSteps
			});
		}
	}

	return solvedBoardsWithSteps;
}

/**
 *
 * @param {number[][][]} matrix
 * @returns
 */
function orchestrateSolution(matrix) {
	const boardSolver = new RuleSolver(matrix);
	const solvedMatrix = boardSolver.solve();
	console.log(boardSolver.state);

	if (boardSolver.state === SolveState.SOLVED) {
		return [[solvedMatrix, boardSolver.state, boardSolver.seqSteps]];
	} else if (boardSolver.state === SolveState.UNSOLVABLE) {
		return [];
	} else {
		const [n, r, c] = findMinimumOptions(solvedMatrix);
		console.log(n, r, c);
		console.log(solvedMatrix[r][c]);
		const possibleSolutions = experimentValues(solvedMatrix, r, c);
		const possibleCompleteSolutions = [];
		for (const [eSolvedMatrix, solutionState, seqSteps] of possibleSolutions) {
			possibleCompleteSolutions.push([
				eSolvedMatrix,
				solutionState,
				boardSolver.seqSteps.concat(seqSteps)
			]);
		}
		return possibleCompleteSolutions;
	}
}

function findMinimumOptions(m) {
	let minCount = 9;
	let r = -1;
	let c = -1;
	for (let i = 0; i < m.length; i++) {
		const row = m[i];
		for (let j = 0; j < row.length; j++) {
			const cell = row[j];
			if (cell.length !== 1 && cell.length < minCount) {
				minCount = cell.length;
				r = i;
				c = j;
			}
		}
	}

	return [minCount, r, c];
}

function experimentValues(matrix, rowIndex, columnIndex) {
	const options = matrix[rowIndex][columnIndex];
	const matrices = [];

	for (const option of options) {
		console.log(`Trying ${option} at (${rowIndex},${columnIndex})`);
		const expStep = [
			[rowIndex + 1, columnIndex + 1],
			option,
			'Experimental'
		];

		const tempM = JSON.parse(JSON.stringify(matrix));
		tempM[rowIndex][columnIndex] = [option];

		const possibleSolutions = orchestrateSolution(tempM);

		for (const [sm, ss, steps] of possibleSolutions) {
			const stepsWithExp = [expStep, ...steps];
			matrices.push([sm, ss, stepsWithExp]);
		}
	}

	const solvedMatrices = matrices.filter(([sm, ss]) => ss === SolveState.SOLVED);

	if (solvedMatrices.length === 0) {
		return [];
	} else {
		return solvedMatrices;
	}
}
